### DISCUSSION TITLE
Obfuscating streamlit application

### DISCUSSION THREAD

**Discussion by jais001** [dashingsoft/pyarmor#1527] [Q&A]
Hi,

I have tried to obfuscate a Streamlit application using the following command:
`pyarmor gen  -O chat  chat-app/`

Obfuscation is a success, but while trying to run the Streamlit application I am getting  the following exception:

```
Uncaught app exception
Traceback (most recent call last):
  File "/home/sfm/Jais/Work/Projects/code obfuscation/mindport-chatbot/venv/lib/python3.11/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 541, in _run_script
    exec(code, module.__dict__)
  File "<frozen __main__>", line 3, in <module>
RuntimeError: unauthorized use of script (1:1075)
```

In a previous discussion I saw someone stating to use `--no-cross-protection`. But it doesn't work with version 8. 
`pyarmor gen --no-cross-protection -O chat  chat-app/`

While running this command I get the following message:
```
pyarmor: error: unrecognized arguments: --no-cross-protection
```

Can we achieve code obfuscation for Streamlit? I would greatly appreciate any guidance or suggestions on how to do this.

**Comment by jais001** [comment]
Hi, I'd greatly appreciate it if someone could offer some guidance on obfuscating Streamlit code. Your input would be very helpful.

**Comment by jondy** [comment]
Try to run `pyarmor cfg restrict_module=0` before obfuscate scripts

**Reply by ayanika02** [reply]
I performed the same command but it did not work. It showed the error ``` RuntimeError: missing license key to run the script (1:10672)```. Suggest the solution please.....

**Comment by madhuri-rai07** [comment]
Hi I am also trying to obfuscate streamlit app but getting following errors. I used `pyarmor gen --platform linux.x86_64 --platform linux.aarch64 -r .` command to generate obfuscated files.

2023-12-01 05:05:15.921 Uncaught app exception
Traceback (most recent call last):
  File "/home/intel/.local/lib/python3.10/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 534, in _run_script
    exec(code, module.__dict__)
  File "<frozen __main__>", line 3, in <module>
RuntimeError: unauthorized use of script (1:1075)
2023-12-01 05:05:20.156 Uncaught app exception
Traceback (most recent call last):
  File "/home/intel/.local/lib/python3.10/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 534, in _run_script
    exec(code, module.__dict__)
  File "<frozen __main__>", line 3, in <module>
RuntimeError: the format of obfuscated script is incorrect (1:1055)

Any help is appreciated

**Comment by LittleTerry** [comment]
same here

**Comment by ahalamora1981** [comment]
Same here.

2024-01-05 08:16:00.429 Uncaught app exception
Traceback (most recent call last):
  File "/mnt/d/Dev/Audio/ASR_AND_TTS/venv-wsl/lib/python3.10/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 534, in _run_script
    exec(code, module.__dict__)
  File "<frozen __main__>", line 3, in <module>
RuntimeError: unauthorized use of script (1:1080)
2024-01-05 08:18:46.937 Uncaught app exception
Traceback (most recent call last):
  File "/mnt/d/Dev/Audio/ASR_AND_TTS/venv-wsl/lib/python3.10/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 534, in _run_script
    exec(code, module.__dict__)
  File "<frozen __main__>", line 3, in <module>
RuntimeError: the format of obfuscated script is incorrect (1:1060)

**Comment by jondy** [comment]
How about disable this check by
```
pyarmor cfg restrict_module=0
```
Then obfuscate the scripts again.

**Reply by LittleTerry** [reply]
I've tried it, but still conldn't work.

**Comment by jondy** [comment]
`RuntimeError: the format of obfuscated script is incorrect (1:1060)`
This errors means the obfuscated scripts may be obfuscated by different licenses, for example, some by trial license, some by Basic license. 
Try to obfuscate all the scripts, also re-generate runtime package.

**Comment by kararsi-ayoub** [comment]
same probleme is there a solution yet ...?

**Comment by john-altitude** [comment]
Same problem, I have a small test case and I am certain every file has been obfuscated using the same trial license. I can run a script using `python3 /path/to/script.py` but if I try to streamlit run it `streamlit run /path/to/script.py` I get:

`RuntimeError: unauthorized use of script (1:1102)`
`RuntimeError: the format of obfuscated script is incorrect (1:1082)`

**Comment by jondy** [comment]
Please refer to
https://pyarmor.readthedocs.io/en/latest/topic/obfuscated-script.html

**Comment by vsilvass** [comment]
Same problem, it happens to me when I use --obf-module 0 instead of the default one.

**Comment by jondy** [comment]
--obf-module 0 results in RuntimeError, it's a bug. It will be fixed in next release v8.5.9

**Comment by Waschbrettwade** [comment]
I have a similar problem. I am using pyarmor 8.5.8 and I am running "pyarmor cfg restrict_module=0" first, and then obfuscate my scripts all at once using "pyarmor gen -r -O dist wrapper.py, dir1, dir2 ...". 
The wrapper looks like this:

```
import streamlit
import streamlit.web.cli as stcli
import os, sys

def resolve_path(path):
    resolved_path = os.path.abspath(os.path.join(os.getcwd(), path))
    return resolved_path

if __name__ == "__main__":
    sys.argv = [
        "streamlit",
        "run",
        resolve_path("utils/app.py"),
        "--global.developmentMode=false",
    ]
    sys.exit(stcli.main())
```

If then, I run the obfuscated wrapper, the Streamlit app runs just fine. 

When using --obf-module=0, the browser doesn't even open, but I immediately get the error that @madhuri-rai07 sent in on Nov 30, 2023, but in a Windows-frame. 

**However**, my goal is to create an EXE (using cx_Freeze or stlite and npm) from my Streamlit app. Whether the EXE is created and then the scripts are obfuscated or the other way around. When running the EXE using obfuscated files, it looks like it's working. But when I click on anything, I get the error:

```
	RuntimeError: the format of obfuscated script is incorrect (1:1082)
	Traceback:

	File "C:\Users\user\Python\Installer cx_freeze_test_2\.venv\Lib\site-packages\streamlit\runtime\scriptrunner\script_runner.py", line 600, in _run_script
	exec(code, module.__dict__)
	File "<frozen __main__>", line 3, in <module>
```

Does anyone know what is causing this and how to fix that behaviour? Thank you very much!

**Comment by TimothyVermeirenBiztory** [comment]
I have tested this with the new version (`v8.5.9`) but the result is the same. I obfuscated a Streamlit app using `pyarmor gen --obf-module 0 app.py`.

The RuntimeError is gone, but we're back to:
```
2024-06-13 13:40:37.100 Uncaught app exception
Traceback (most recent call last):
File "/usr/local/lib/python3.9/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 600, in _run_script
   exec(code, module.__dict__)
File "<frozen __main__>", line 3, in <module>
RuntimeError: unauthorized use of script (1:1102)
```

**However! A rectification...** I was assuming that `--obf-module` would be equivalent to setting `pyarmor cfg restrict_module=0`. It seems like that was an incorrect assumption. I have actually been able to make this work with:

```
pyarmor cfg restrict_module=0
pyarmor gen --mix-str app.py
```

**But then!** While the script does start the web app and it _looks_ like it's working, I then get the exact same behavior as described by @Waschbrettwade in the previous post. As soon as you interact:

```
RuntimeError: the format of obfuscated script is incorrect (1:1082)
Traceback:
File "/usr/local/lib/python3.9/site-packages/streamlit/runtime/scriptrunner/script_runner.py", line 600, in _run_script
    exec(code, module.__dict__)
File "<frozen __main__>", line 3, in <module>
```

**Comment by jondy** [comment]
Try disable this option:
```
pyarmor cfg clear_module_co = 0
```
It may fix this problem `RuntimeError: the format of obfuscated script is incorrect (1:1082)
`

**Reply by TimothyVermeirenBiztory** [reply]
Appreciate the quick response, @jondy, thank you! Disabling `clear_module_co` does change something, but I'm not sure how to interpret what's happening next.

At that point, when interacting with the Streamlit app, instead of showing the previous `RuntimeError`, the app now seems to "crash" and restart:

![CleanShot 2024-06-14 at 13 08 49](https://github.com/dashingsoft/pyarmor/assets/57324589/b517eaff-cb6c-43cd-a4f6-e6d907b8b58c)

That would be on the lines where we see `<applicationname_redacted>-lite-1 exited with code 0`. Nowhere in my own code is the application supposed to exit with code `0`, and in the browser the screen just shows the application being inactive and reloading from start after a few seconds.

I should probably have noted, and you can tell from the screenshot: in this context the application is being run in a Docker container. If I obfuscate all files _except_ the one that the Streamlit app is in, everything works fine.

I'm not sure if that is very useful information, but I'm happy to look up more details if it can help.

**Comment by jondy** [comment]
I'm not sure pyarmor could work with streamlit because I don't know well about streamlit.
From traceback I found streamlit is using `exec(co, module_dict)` to execute the module, and may execute same co twice, so disable `clear_module_co` only fixs loading twice issues. 
Because streamlit is not `import` module, but use and patch code object directly, so I think streamlit may not work with Pyarmor well.

**Comment by TimothyVermeirenBiztory** [comment]
Noted, thank you @jondy.

**Comment by Hu-Wentao** [comment]
any update?

**Comment by jondy** [comment]
I have explain it
> Because streamlit is not `import` module, but use and patch code object directly, so I think streamlit may not work with Pyarmor well.
Pyarmor goal is protecting code object couldn't read by others, if streamlit need patch code object and use it directly (not run code by Python interpreter), I think it doesn't work with Pyarmor.

I'm not sure `--obf-code 0` which not obfuscate code object works with streamit or not.

**Comment by Hu-Wentao** [comment]
Thanks to @jondy  for the answer.
I'm trying to wrap streamlit with [nuitka](https://github.com/Nuitka/Nuitka), hiding the source code as an extra file in the exe, which will be unpacked in the system temporary directory when running. 
This seems to work in principle, and if anyone else is trying it, welcome to discuss.

**Reply by Waschbrettwade** [reply]
So, do you mean, when running the exe the source code will be readable temporarily?
